src/coprocessor/dag/expr/builtin_cast.rs

```diff
@@ -172,299 +171,497 @@ impl FnCall {
+            let j: Json = try!(s.parse());
+            Ok(Some(Cow::Owned(j)))
+        } else {
+            Ok(Some(Cow::Owned(Json::String(s))))
+        }
     }
 
-    pub fn cast_time_as_json(
-        &self,
+    pub fn cast_time_as_json<'a, 'b: 'a>(
+        &'b self,
         ctx: &StatementContext,
-        row: &[Datum],
-    ) -> Result<Option<Vec<Json>>> {
-        unimplemented!()
+        row: &'a [Datum],
+    ) -> Result<Option<Cow<'a, Json>>> {
+        let val = try_opt!(self.children[0].eval_time(ctx, row));
+        let mut val = val.into_owned();
+        if val.get_tp() == types::DATETIME || val.get_tp() == types::TIMESTAMP {
+            val.set_fsp(mysql::MAX_FSP as u8);
+        }
+        let s = format!("{}", val);
+        Ok(Some(Cow::Owned(Json::String(s))))
+    }
+
+    pub fn cast_duration_as_json<'a, 'b: 'a>(
+        &'b self,
+        ctx: &StatementContext,
+        row: &'a [Datum],
+    ) -> Result<Option<Cow<'a, Json>>> {
+        let val = try_opt!(self.children[0].eval_duration(ctx, row));
+        let mut val = val.into_owned();
+        val.fsp = mysql::MAX_FSP as u8;
+        let s = format!("{}", val);
+        Ok(Some(Cow::Owned(Json::String(s))))
     }
 
-    pub fn cast_duration_as_json(
+    pub fn cast_json_as_json<'a, 'b: 'a>(
+        &'b self,
+        ctx: &StatementContext,
+        row: &'a [Datum],
+    ) -> Result<Option<Cow<'a, Json>>> {
+        self.children[0].eval_json(ctx, row)
+    }
+
+    fn produce_dec_with_specified_tp(
         &self,
         ctx: &StatementContext,
-        row: &[Datum],
-    ) -> Result<Option<Vec<Json>>> {
-        unimplemented!()
+        val: Decimal,
+    ) -> Result<Cow<Decimal>> {
+        let flen = self.tp.get_flen();
+        let decimal = self.tp.get_decimal();
+        if flen == convert::UNSPECIFIED_LENGTH || decimal == convert::UNSPECIFIED_LENGTH {
+            return Ok(Cow::Owned(val));
+        }
+        let res = try!(val.convert_to(ctx, flen as u8, decimal as u8));
+        Ok(Cow::Owned(res))
     }
 
-    pub fn cast_json_as_json(
+    /// `produce_str_with_specified_tp`(`ProduceStrWithSpecifiedTp` in tidb) produces
+    /// a new string according to `flen` and `chs`.
+    fn produce_str_with_specified_tp(
         &self,
         ctx: &StatementContext,
-        row: &[Datum],
-    ) -> Result<Option<Vec<Json>>> {
-        unimplemented!()
+        mut s: Vec<u8>,
+    ) -> Result<Cow<Vec<u8>>> {
+        let flen = self.tp.get_flen();
+        let chs = self.tp.get_charset();
+        if flen < 0 {
+            return Ok(Cow::Owned(s));
+        }
+        let flen = flen as usize;
+        // flen is the char length, not byte length, for UTF8 charset, we need to calculate the
+        // char count and truncate to flen chars if it is too long.
+        if chs == charset::CHARSET_UTF8 || chs == charset::CHARSET_UTF8MB4 {
+            let s = try!(String::from_utf8(s));
+            let char_count = s.char_indices().count();
```

Comment:
Merge would be failed here since count would move the iterator https://doc.rust-lang.org/std/iter/trait.Iterator.html#method.count @BusyJay